doc/proposals/tls-utilities.md

```diff
@@ -0,0 +1,159 @@
+
+Each user specified CRD represents a self contained system. Based on this, we can assume that any TLS assets generated are specific to each system which then is also specific to the corresponding CRD.
+
+Based on the above assumption, we want to design a TLS utility pkg that helps generate TLS assets and ties them to a specific CRD.
+
+## Goals:
+
+To provide TLS utilities for operator developers to create self signed Certificate Authority (CA) and the TLS encryption keys along with the signed certs. This will enable the operator to setup TLS assets for the application so that all the communication within it is secure.
+
+## Details:
+
+We will break down the TLS workflow into the following steps:
+
+1. Generate a Certificate Authority (CA): A CA is a centralized trusted third party that signs the certificates. We need to generate a CA key and CA Certificate to get started. If the users want to provide their own CA, we can add an option to do so. The generated CA certificate can be saved in a ConfigMap as follows:
+	
+	Inputs:
+
+	* CRD-kind
+	* CRD-name
+	* namespace
+
+	Outputs:
+
+        * A ConfigMap that contains the CA Cert:
+
+	```
+	kind: ConfigMap
+	apiVersion: v1
+	metadata:
+  	name: <crd-kind>-<crd-name>-ca
+  	namespace: <ns>
+	data:
+  	ca.pem: ...
+            ```
+        * A secret that contains the CA key:
+
+           ```
+ 	kind: Secret
+ 	apiVersion: v1
+ 	metadata:
+  	name: <cr-kind>-<cr-name>-ca
+  	namespace: <ns>
+ 	data:
+  	ca-key.pem: ...
+	```
+
+2. Verify the CA certificate: The newly generated CA certificate in ConfigMap should be validated i.e. the digital signature should be checked, the expiry, activation dates and validity period should be checked, etc.
+
+3. Generate Server or Client Certificate: We assume that the components inside an application communicate through a service. Therefore, we generate a TLS private key and certificate for that service.
+
+	Inputs:
+	
+	* CRD-kind
+	* CRD-name
+	* Namespace
+	* Previously generated CA key and Certificate
+
+	Options:
+
+	* Key usage: [Server, Client or Both] If Server, needs svc obj (cluster.local can be different). If client, needs CN and possible org.
+	* Option to specify signature configuration policy i.e Signing Profile for the certificate.
+
+        Outputs:
+
+	* A Secret containing Private Key and Server/Client Certificate signed by the CA:
+
+	```
+	kind: Secret
+	apiVersion: v1
+	metadata:
+	 name: <cr-kind>-<cr-name>-<CertName>
+	 namespace: <ns>
+	data:
+	 tls.crt: ...
+	 tls.key: …
+	type: kubernetes.io/tls
+	```
+
+4. Verify Server or Client Certificate: The newly generated client/server certificate in should be validated i.e. the digital signature should be checked, the expiry, activation dates and validity period should be checked, certificate chain should be verified etc
```

Comment:
`The newly generated client/server certificate in should be validated`
Why do we need to do that? the key and cert generation mechanism should generate a proper cert right?

Reply:
Ideally it should, but it is a common openssl practice to re-check and validate it.